core/src/main/java/jenkins/model/Nodes.java

```diff
@@ -139,6 +140,7 @@ public void run() {
             });
             // TODO there is a theoretical race whereby the node instance is updated/removed after lock release
             persistNode(node);
+            NodeListener.fireOnCreated(node);
```

Comment:
To put it another way. I have concerns that listeners may do too much work as to cause delays elsewhere that cause excessive provisioning. IMHO it is not sufficient to notify outside of the lock, we should rather notify outside of the action thread.

Based on the use cases for this listener that may affect the listener API (eg if you want to capture the state change and it changes again before the listener is called)

There are also concerns about a slow listener backing up and perhaps being multiple seconds behind the current state... That can be addressed in one of two ways:

1. Remove in order delivery guarantee and allow concurrent calls to the listener (the listener can then do nothing to reconstruct the order as thread scheduling is not guaranteed to be in order so even making your listener `synchronized` will not protect you - this is why in order delivery *per listener* needs to be part of the listener API)
2. Allow dropping events on the floor if the backlog is too much (perhaps giving the listener a "you're too slow" event)

OTOH we can perhaps ignore that concern for now until we get an implementation that is bad... But I think I would favour adding a "eventsLostBecauseYouAreTooSlow" method to the listener API up front, even if initially we do not call it, so that anyone implementing is aware that we may add that behaviour to the contract